src/server/accept.rs

```diff
@@ -172,14 +159,20 @@ fn handle_error(error: ServerError) {
     }
 }
 
-async fn drain_connections(connections: &mut JoinSet<()>) {
+async fn drain_connections(immediate: bool, mut connections: JoinSet<Result<(), ServerError>>) {
     if cfg!(debug_assertions) {
-        println!("draining {} inflight connections...", connections.len());
+        println!("joining {} inflight connections...", connections.len());
     }
 
     while let Some(result) = connections.join_next().await {
-        if let Err(error) = &result {
-            handle_error(ServerError::Join(error));
+        match result {
+            Ok(Err(error)) => handle_error(error),
+            Err(error) => handle_error(ServerError::Join(error)),
+            _ => {}
+        }
+
+        if !immediate {
+            coop::consume_budget().await;
```

Comment:
When we're joining in the background, yield to the runtime if we exhausted our budget.